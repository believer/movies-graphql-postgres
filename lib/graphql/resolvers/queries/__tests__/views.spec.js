const views = require('../views')
const { verifyToken } = require('../../../../services/token')

jest.mock('../../../../services/token', () => ({
  verifyToken: jest.fn().mockImplementation(() =>
    Promise.resolve({
      id: 2,
    })
  ),
}))

describe('queries/views', () => {
  let db

  beforeEach(() => {
    db = {
      query: jest.fn().mockReturnValue('dbData'),
    }
  })

  it('should verify token', () => {
    views(null, {}, { db, token: 'test' })

    expect(verifyToken).toHaveBeenCalledWith('test')
  })

  it('should get views', async () => {
    await views(null, { offset: 0, limit: 10 }, { db })

    expect(db.query.mock.calls[0][0]).toMatchSnapshot()
  })

  it('should return database data', async () => {
    const response = await views(null, { offset: 0, limit: 10 }, { db })

    expect(response).toEqual('dbData')
  })

  it('should handle errors', async () => {
    db.query.mockImplementation(() => Promise.reject('nope'))

    try {
      await views(null, { offset: 0, limit: 10 }, { db })
    } catch (e) {
      expect(e.message).toMatch(/nope/)
    }
  })
})
