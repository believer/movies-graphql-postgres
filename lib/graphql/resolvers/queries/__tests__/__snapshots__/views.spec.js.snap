// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queries/views should get views 1`] = `
"
      SELECT
        seen.id as movie_id,
        seen.title,
        json_agg(json_build_object('date', seen.view_date, 'userId', seen.user_id)) as dates,
        COUNT(seen.view_date) as views_count
      FROM (
        SELECT 
          m.id,
          m.title,
          v.view_date,
          v.user_id
        FROM 
          movies AS m
        INNER JOIN views AS v ON v.movie_id = m.id
        WHERE v.user_id = 2
        ORDER BY m.id, v.view_date DESC
      ) as seen
      GROUP BY id, title
      ORDER BY views_count DESC, title ASC
      OFFSET 0
      LIMIT 10
    "
`;
